openmoltools/packmol.py

```diff
@@ -121,52 +122,60 @@ def pack_box(pdb_filenames_or_trajectories, n_molecules_list, tolerance=2.0, box
 
     """
     assert len(pdb_filenames_or_trajectories) == len(n_molecules_list), "Must input same number of pdb filenames as num molecules"
-    
-    pdb_filenames = []
-    for obj in pdb_filenames_or_trajectories:
-        try:  # See if MDTraj Trajectory
-            tmp_filename = tempfile.mktemp(suffix=".pdb")
-            obj.save_pdb(tmp_filename)
-            pdb_filenames.append(tmp_filename)
-        except AttributeError:  # Not an MDTraj Trajectory, assume filename
-            pdb_filenames.append(obj)
-    
+
     if PACKMOL_PATH is None:
         raise(IOError("Packmol not found, cannot run pack_box()"))
-    
-    output_filename = tempfile.mktemp(suffix=".pdb")
-
-    # approximating volume to initialize  box
-    if box_size is None:
-        box_size = approximate_volume(pdb_filenames, n_molecules_list)    
 
-    header = HEADER_TEMPLATE % (tolerance, output_filename)
-    for k in range(len(pdb_filenames)):
-        filename = pdb_filenames[k]
-        n_molecules = n_molecules_list[k]
-        header = header + BOX_TEMPLATE % (filename, n_molecules, box_size, box_size, box_size)
-    
-    pwd = os.getcwd()
-    
-    print(header)
-    
-    packmol_filename = "packmol_input.txt"
-    packmol_filename = tempfile.mktemp(suffix=".txt")
-    
-    file_handle = open(packmol_filename, 'w')
-    file_handle.write(header)
-    file_handle.close()
-    
-    print(header)
-
-    os.system("%s < %s" % (PACKMOL_PATH, packmol_filename)) 
+    pdb_filenames = []
 
-    trj = md.load(output_filename)
+    # We save all the temporary files in a temporary directory
+    # that is deleted at the end of the function.
+    with temporary_directory() as tmp_dir:
+
+        # Convert all mdtraj.Trajectory objects to PDB files.
+        for obj in pdb_filenames_or_trajectories:
+            try:  # See if MDTraj Trajectory
+                tmp_filename = tempfile.mktemp(suffix=".pdb", dir=tmp_dir)
+                obj.save_pdb(tmp_filename)
+                pdb_filenames.append(tmp_filename)
+            except AttributeError:  # Not an MDTraj Trajectory, assume filename
+                pdb_filenames.append(obj)
+
+        # Approximating volume to initialize box.
+        if box_size is None:
+            box_size = approximate_volume(pdb_filenames, n_molecules_list)
+
+        # Adjust box_size for periodic box. Packmol does not explicitly
+        # support periodic boundary conditions and the suggestion on
+        # theri docs is to pack in a box 2 angstroms smaller. See
+        # http://www.ime.unicamp.br/~martinez/packmol/userguide.shtml#pbc
+        packmol_box_size = box_size - 2  # angstroms
```

Comment:
I think we're both saying variants of the same thing: Your current approach results in using a box size that is 2 angstroms smaller than intended throughout, which is not quite what we want. 

I think you either want:
a) John's approach: Pack into the target box size, then increase the box size by 2 angstroms after packing into it
b) My approach: Pack into a box size 2 angstroms smaller than the target, then increase the box size by 2 angstroms after packing into it 

Either can work. But approach (a) will require you changing the behavior of `pack_box` when `box_size` is provided (and updating doc strings) so that if the user asks for a 4 nm box, you don't give them back a 4.2 nm box (which is what approach (a) would do). Approach (b) would leave the behavior of `pack_box` the same when the user provides `box_size` (by packing into a box 2 angstroms smaller than expanding it by 2 angstroms) and would fix the "overly dense box" issue by modifying the functions `approximate_volume...` to give box edge estimates which are 2 angstroms larger than they need to be, so you have room to (i) shrink by 2 angstroms, (ii) pack, and (iii) increase by 2 angstroms.

So basically I think you need to do (a) or (b) and I'm OK with either. But your current approach will probably fail for boxes which are too dense, AND results in final box sizes which still don't include the 2 angstrom buffer unless I'm missing where that's done.